File: test/language/union_types/java/union_types/UnionWithArrayTest.java
package union_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;

import zserio.runtime.io.SerializeUtil;

import union_types.union_with_array.TestUnion;
import union_types.union_with_array.Data8;

public class UnionWithArrayTest
{
    @Test
    public void array8()
    {
        final TestUnion test = new TestUnion();
        test.setArray8(new Data8[4]);
        assertEquals(4, test.getArray8().length);
    }

    @Test
    public void array16()
    {
        final TestUnion test = new TestUnion();
        test.setArray16(new short[4]);
        assertEquals(4, test.getArray16().length);
    }

    @Test
    public void writeReadFile() throws IOException
    {
        {
            final TestUnion testUnion = new TestUnion();
            testUnion.setArray8(createArray8());
            final File file = new File(BLOB_NAME_BASE + "array8.blob");
            SerializeUtil.serializeToFile(testUnion, file);

            final TestUnion readTestUnion = SerializeUtil.deserializeFromFile(TestUnion.class, file);
            assertEquals(testUnion, readTestUnion);
        }

        {
            final TestUnion testUnion = new TestUnion();
            testUnion.setArray16(createArray16());
            final File file = new File(BLOB_NAME_BASE + "array16.blob");
            SerializeUtil.serializeToFile(testUnion, file);

            final TestUnion readTestUnion = SerializeUtil.deserializeFromFile(TestUnion.class, file);
            assertEquals(testUnion, readTestUnion);
        }
    }

    @Test
    public void hashCodeMethod()
    {
        TestUnion testUnion1 = new TestUnion();
        TestUnion testUnion2 = new TestUnion();
        assertEquals(testUnion1.hashCode(), testUnion2.hashCode());
        testUnion1.setArray8(createArray8());
        assertFalse(testUnion1.hashCode() == testUnion2.hashCode());
        testUnion2.setArray8(createArray8());
        assertEquals(testUnion1.hashCode(), testUnion2.hashCode());
        testUnion2.getArray8()[0].setData((byte)0);
        assertFalse(testUnion1.hashCode() == testUnion2.hashCode());
        testUnion2.setArray16(createArray16());
        assertFalse(testUnion1.hashCode() == testUnion2.hashCode());

        // use hardcoded values to check that the hash code is stable
        assertEquals(87386744, testUnion1.hashCode());
        assertEquals(1575145265, testUnion2.hashCode());

        testUnion1.setArray16(createArray16());
        assertEquals(testUnion1.hashCode(), testUnion2.hashCode());
    }

    private static Data8[] createArray8()
    {
        return new Data8[]{new Data8((byte)-1), new Data8((byte)-2), new Data8((byte)-3), new Data8((byte)-4)};
    }

    private static short[] createArray16()
    {
        return new short[]{-10, -20, -30, -40, -50};
    }

    private static final String BLOB_NAME_BASE = "union_with_array_";
}

File: test/language/union_types/python/UnionWithArrayTest.py
import os
import zserio

import UnionTypes
from testutils import getApiDir

class UnionWithArrayTest(UnionTypes.TestCase):
    def testArray8(self):
        testUnion = self.api.TestUnion()
        testUnion.array8 = [self.api.Data8(), self.api.Data8(), self.api.Data8(), self.api.Data8()]
        self.assertEqual(4, len(testUnion.array8))

    def testArray16(self):
        testUnion = self.api.TestUnion(array16_=[1, 2, 3, 4])
        self.assertEqual(4, len(testUnion.array16))

    def testWriteReadFileArray8(self):
        testUnion = self.api.TestUnion(array8_=self._createArray8())
        filename = self.BLOB_NAME_BASE + "array8.blob"
        zserio.serialize_to_file(testUnion, filename)

        readTestUnion = zserio.deserialize_from_file(self.api.TestUnion, filename)
        self.assertEqual(testUnion, readTestUnion)

    def testWriteReadFileArray16(self):
        testUnion = self.api.TestUnion(array16_=self._createArray16())
        filename = self.BLOB_NAME_BASE + "array16.blob"
        zserio.serialize_to_file(testUnion, filename)

        readTestUnion = zserio.deserialize_from_file(self.api.TestUnion, filename)
        self.assertEqual(testUnion, readTestUnion)

    def testHash(self):
        testUnion1 = self.api.TestUnion()
        testUnion2 = self.api.TestUnion()
        self.assertEqual(hash(testUnion1), hash(testUnion2))
        testUnion1.array8 = self._createArray8()
        self.assertNotEqual(hash(testUnion1), hash(testUnion2))
        testUnion2.array8 = self._createArray8()
        self.assertEqual(hash(testUnion1), hash(testUnion2))
        testUnion2.array8[0].data = 0
        self.assertNotEqual(hash(testUnion1), hash(testUnion2))
        testUnion2.array16 = self._createArray16()
        self.assertNotEqual(hash(testUnion1), hash(testUnion2))

        # use hardcoded values to check that the hash code is stable
        # using __hash__ to prevent 32-bit Python hash() truncation
        self.assertEqual(87386744, testUnion1.__hash__())
        self.assertEqual(1575145265, testUnion2.__hash__())

        testUnion1.array16 = self._createArray16()
        self.assertEqual(hash(testUnion1), hash(testUnion2))

    def _createArray8(self):
        return [self.api.Data8(-1), self.api.Data8(-2), self.api.Data8(-3), self.api.Data8(-4)]

    @staticmethod
    def _createArray16():
        return [-10, -20, -30, -40, -50]

    BLOB_NAME_BASE = os.path.join(getApiDir(os.path.dirname(__file__)), "union_with_array_")
